sound/soc/intel/boards/sof_sdw.c

```diff
@@ -1311,8 +1323,20 @@ static int sof_card_dai_links_create(struct device *dev,
 	for (i = 0; i < SDW_MAX_GROUPS; i++)
 		group_generated[i] = false;
 
-	/* generate DAI links by each sdw link */
 	for (; adr_link->num_adr; adr_link++) {
+		/*
+		 * If there is another codec on the same sdw link, we have to append the
+		 * codec type to the dai link name to prevent duplicated dai link name
```

Comment:
that wouldn't work for the two amplifier on the same link case, would it?

I think you need to test if the part ID and manufacturer ID are different.

Reply:
@plbossart I tested on a chrome device with dual max98373 on the same sdw link and it works.

Reply:
I don't understand your explanations @bardliao. What do you mean by "we won't see another adr_link with the same mask". 

We currently only have a notion of "link_mask" which describe which links are useful/enabled on a given platform. That's used to match with what the platform supports. And then for each adr_link, we will have the list of all devices connected. They could be identical or they could be different. It doesn't matter, they will be on one adr_link.

Maybe you can share what you were planning for the rt712 ACPI tables, that would help me understand what you are suggesting.

Reply:
```
static const struct snd_soc_acpi_adr_device rt712_0_single_adr[] = {
	{
		.adr = 0x000030025D071201ull,
		.num_endpoints = 1,
		.endpoints = &single_endpoint,
		.name_prefix = "rt712"
	}
};

static const struct snd_soc_acpi_adr_device rt1712_0_single_adr[] = {
	{
		.adr = 0x000030025D171201ull,
		.num_endpoints = 1,
		.endpoints = &single_endpoint,
		.name_prefix = "rt712-dmic"
	}
};

static const struct snd_soc_acpi_link_adr tgl_712_only[] = {
	{
		.mask = BIT(0),
		.num_adr = ARRAY_SIZE(rt712_0_single_adr),
		.adr_d = rt712_0_single_adr,
	},
	{
		.mask = BIT(0),
		.num_adr = ARRAY_SIZE(rt1712_0_single_adr),
		.adr_d = rt1712_0_single_adr,
	},
	{}
};
```

sof_sdw assumes devices on the same sdw link are all the same. We will need to some more changes on sdw_sof.c if we want to put different devices in a snd_soc_acpi_adr_device array.

Reply:
Ah yes, it looks like we have restrictions indeed. I don't think it's quite right to repeat the BIT(0) mask with individual devices. That was not the intent, what I really wanted is one entry per link, and all the devices on that link listed.

That's what I had in mind - different level to add the devices

````c
static const struct snd_soc_acpi_adr_device rt712_0_adr[] = {
	{
		.adr = 0x000030025D071201ull,
		.num_endpoints = 1,
		.endpoints = &single_endpoint,
		.name_prefix = "rt712"
	},
	{
		.adr = 0x000030025D171201ull,
		.num_endpoints = 1,
		.endpoints = &single_endpoint,
		.name_prefix = "rt712-dmic"
	}

};

static const struct snd_soc_acpi_link_adr tgl_712_only[] = {
	{
		.mask = BIT(0),
		.num_adr = ARRAY_SIZE(rt712_0_adr),
		.adr_d = rt712_0_adr,
	},
	{}
};

````

Reply:
@plbossart I updated the code, and we will need some surgery on sof_sdw machine driver to remove the restrictions.